### PR TITLE
Fix protocol evaluation

### LINKED ISSUE #11656: Self-Hosted Studio makes requests to `localhost:8000` instead of `EXTERNAL_API_URL`
# Bug report

## Describe the bug

When loading tables and buckets, the `Studio` makes requests to `http://localhost:8000/storage/v1/bucket` and `http://localhost:8000/rest/v1/` instead of `https://kong.example.com/rest/v1/`. This means that if I visit the `Studio` externally it will be stuck on `Connecting to default project` because only the host computer can reach `localhost`.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Clone the master branch
2. Create subdomains that point to point `8000` and `3000`
3. Use the subdomain that points to `8000` as `API_EXTERNAL_URL` and `SUPABASE_PUBLIC_URL`  in `.env`
4. `docker compose up -d`
5. Visit the `3000` subdomain on the host computer and it works but you will see in `Network` that the requests are made to `localhost:8000`
6. Visit the `3000` subdomain on a different computer and it does not work,
    - You will see in `Network` that the requests are made to `localhost:8000`
    - External computer cannot reach that
    - Everthing will show `Connecting to default project`

## Expected behavior

The `Studio` should use the correct URL to make requests.

## Screenshots

image

## Additional context

I know that this is a bug and not an issue with my configuration because it worked before AND my configuration still works when I run the `Studio` locally. According to [a comment on a separate bug fix](https://github.com/supabase/supabase/pull/11331#issuecomment-1375561791), this bug was introduced a few weeks ago on the docker image.

### PR DESCRIPTION
**Pull request by ruggi99**
## What kind of change does this PR introduce?

Bug fix

## What is the current behavior?

Closes #11656

## What is the new behavior?

Actually, when self-hosting Supabase, we can change `SUPABASE_PUBLIC_URL` to tell Supabase where to find Kong and all other services. The fact is that the address is not taken in consideration due to some bugs.
First the protocol to be used is evaluated at build time, so when the Docker image is built, not good.
Second the endpoint is also evaluated at buildtime, but this is due to this PR #11331.
The commits that introduced this bug are: https://github.com/supabase/supabase/commit/0868ee96f27333cd541dd5d512368a584a510210 and https://github.com/supabase/supabase/commit/3070064cc54b82a84c5eff4810170e40277e772d
Those commit are simply wrong because they are trying to pull serverside data into clientside at runtime and it doesn't work like this, this data is "translated" at builtime.
Instead the correct pattern is to pass down the protocol via props like proposed in this PR.
I thought we can evaluate the protocol based on the protocol on the frontend, but it won't work 100%.

## Additional context

Add any other context or screenshots.

### CODE REVIEW

**Comment by ruggi99** [comment]
I've built the studio image locally, changed docker-compose and can confirm it works as expected

**Comment by ruggi99** [comment]
Is there something blocking this PR?

File: studio/.env
Comment (sweatybridge) [inline studio/.env:10]: Do we know if it's safe to remove these variables? Could some clients depend on this because it's public?

May be we can remove in a separate PR just in case.

File: studio/pages/api/props/project/[ref]/settings.ts
Comment (sweatybridge) [inline studio/pages/api/props/project/[ref]/settings.ts:62]: Hmm we try to keep this static api response symmetric with what our platform api returns. This change would make the static api response different from that of platform...

File: studio/localStores/storageExplorer/StorageExplorerStore.js
Comment (sweatybridge) [inline studio/localStores/storageExplorer/StorageExplorerStore.js:99]: Is it possible to default to https?

**Review by sweatybridge — COMMENTED** [review]
I took a brief look. Main concern is the static api response change. Will need to think about it more before committing

File: studio/.env
Comment (ruggi99) [reply studio/.env:10]: This is a revert from a PR made about 3 weeks ago. I don't think the variables are used, but I can split in a separate PR or eventually leave them there, they are not conflicting AFAIK

File: studio/pages/api/props/project/[ref]/settings.ts
Comment (ruggi99) [reply studio/pages/api/props/project/[ref]/settings.ts:62]: Yes I know, but the protocol is ignored when using the platform.
Maybe should I add a comment explaining why it is needed?

File: studio/localStores/storageExplorer/StorageExplorerStore.js
Comment (ruggi99) [reply studio/localStores/storageExplorer/StorageExplorerStore.js:99]: I set to `http` due to the two commits I linked in the PR template, but setting to `https` should be safe.
Actually that argument should never be undefined in the code and so the default value should never be used.
I added it to not break the API of that function

File: studio/pages/api/props/project/[ref]/settings.ts
Comment (sweatybridge) [reply studio/pages/api/props/project/[ref]/settings.ts:62]: Yup, let's add a comment for now and proceed with this change.

```suggestion
          // TODO: update platform api to return this field
          protocol: PROJECT_ENDPOINT_PROTOCOL
```

@inian do you have any strong opinions about adding the protocol field to middleware api?

File: studio/localStores/storageExplorer/StorageExplorerStore.js
Comment (sweatybridge) [reply studio/localStores/storageExplorer/StorageExplorerStore.js:99]: How about setting the default to `PROJECT_ENDPOINT_PROTOCOL` to be consistent with current behaviour if user pass in `undefined`?

```suggestion
  initStore(projectRef, url, serviceKey, protocol = PROJECT_ENDPOINT_PROTOCOL) {
```

File: studio/components/layouts/StorageLayout/StorageLayout.tsx
Comment (inian) [inline studio/components/layouts/StorageLayout/StorageLayout.tsx:45]: This will be undefined in our production API. 

My preference is to retain the same API format for both self hosted and prod. With this PR, we add a field to only the self hosted API and not the production one which would lead to bugs only in one environment (like this one being undefined in production). Not opposed to adding a protocol field, but lets do it internally too.

File: studio/localStores/storageExplorer/StorageExplorerStore.js
Comment (inian) [inline studio/localStores/storageExplorer/StorageExplorerStore.js:99]: wont `PROJECT_ENDPOINT_PROTOCOL` be undefined here?

**Review by inian — CHANGES_REQUESTED** [review]

File: studio/components/layouts/StorageLayout/StorageLayout.tsx
Comment (ruggi99) [reply studio/components/layouts/StorageLayout/StorageLayout.tsx:45]: Yes we know it will be undefined in production, but also it's always bypassed in production.
It doesn't lead to bugs for now but it'll be in the future due to the mismatch of the API (for example someone modifies code without knowing the mismatch).
So for now it is safe.
Also if you add the protocol field it is safe to remove all conditionals, for example: `{IS PLATFORM ? "https" : protocol}`

File: studio/localStores/storageExplorer/StorageExplorerStore.js
Comment (ruggi99) [reply studio/localStores/storageExplorer/StorageExplorerStore.js:99]: AFAIK `PROJECT_ENDPOINT_PROTOCOL` will be always "http"

Comment (inian) [reply studio/localStores/storageExplorer/StorageExplorerStore.js:99]: it is not imported in this file, will lead to an error

File: studio/components/layouts/StorageLayout/StorageLayout.tsx
Comment (inian) [reply studio/components/layouts/StorageLayout/StorageLayout.tsx:45]: yeah that would be ideal @sweatybridge - adding it to the prod api, wdyt?

File: studio/localStores/storageExplorer/StorageExplorerStore.js
Comment (ruggi99) [reply studio/localStores/storageExplorer/StorageExplorerStore.js:99]: You're right.
Rethinking at it, since `PROJECT_ENDPOINT_PROTOCOL` is always "http" because it is client-side code, maybe I should revert the commit suggestion?

File: studio/components/layouts/StorageLayout/StorageLayout.tsx
Comment (sweatybridge) [reply studio/components/layouts/StorageLayout/StorageLayout.tsx:45]: Sounds good. I will update our api next week

File: studio/localStores/storageExplorer/StorageExplorerStore.js
Comment (sweatybridge) [inline studio/localStores/storageExplorer/StorageExplorerStore.js:11]: @ruggi99 It was imported before https://github.com/supabase/supabase/pull/11740#discussion_r1083419924

You can revert this delete.

Comment (ruggi99) [reply studio/localStores/storageExplorer/StorageExplorerStore.js:11]: Yeah I know, but the whole point is that `PROJECT_ENDPOINT_PROTOCOL` is always "http" because it is client-side code.
But I will revert it so we can move forward

Comment (sweatybridge) [reply studio/localStores/storageExplorer/StorageExplorerStore.js]: Thanks for helping with this change. Although the value is always http, it's a good practice to move string constants to one place instead of having them scattered in different files.

Comment (ruggi99) [reply studio/localStores/storageExplorer/StorageExplorerStore.js]: Sounds good from this perspective

**Review by sweatybridge — DISMISSED** [review]

**Comment by ruggi99** [comment]
@sweatybridge what's the plan here?
You gonna merge this and then merge yours on top of this?

**Comment by sweatybridge** [comment]
If you don't mind, can we close this PR to focus on the new one instead? The platform api has been updated to return `protocol: "https"` field as part of autoApiService object. We didn't want to pollute `app_config` with fields that are not actually configurable.

```json
{
  ...
  "autoApiService": {
    "id": 234857,
    "name": "Default API",
    "app_config": {
      "endpoint": "...",
      "db_schema": "public",
      "realtime_multitenant_enabled": true
    },
    "app": { "id": 1, "name": "Auto API" },
    "protocol": "https",
    "endpoint": "...",
    ...
  }
}
```

**Comment by ruggi99** [comment]
Actually I saw that you used my commits in your PR, so I'll be happy if you merge this to give me "credit".
Your PR will be rebased automatically if I'm not wrong.
Also, since this PR fixes self-host side (platform is untouched) as a temp fix and you have opened a PR with a definitive fix I suggest to merge this asap. I think there are people (who self-host) waiting for this fix.

What do you think?

**Comment by sweatybridge** [comment]
I've rebased my PR on top of master so that your commit history is preserved.

To unblock, I've also triggered a [manual build](https://github.com/supabase/supabase/actions/runs/4016542892) of studio image. Once that's published, you should be able to pull with docker tag `supabase/studio:20230126-6e037f1`.

**Comment by ruggi99** [comment]
Ok thanks, will PR for updating `docker-compose.yml`

**Pull request closed without merging**
